src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -596,17 +610,73 @@ public override void Dispose()
 
         private void Dispose(bool disposing)
         {
-            if (_disposed)
+            int disposed = Interlocked.Exchange(ref _disposed, 1);
+            if (disposed != 0)
             {
                 return;
             }
 
-            _disposed = true;
-            _state.Handle.Dispose();
-            Marshal.FreeHGlobal(_state.SendQuicBuffers);
-            if (_stateHandle.IsAllocated) _stateHandle.Free();
-            CleanupSendState(_state);
-            _state.ConnectionState?.RemoveStream(this);
+            bool callShutdown = false;
+            bool abortRead = false;
+            bool releaseHandles = false;
+            lock (_state)
+            {
+                if (_canWrite && _state.ShutdownWriteState == ShutdownWriteState.None)
+                {
+                    callShutdown = true;
+                }
+
+                if (_canRead && _state.ReadState != ReadState.Aborted && _state.ReadState != ReadState.ReadsCompleted)
+                {
+                    abortRead = true;
+                    _state.ReadState = ReadState.Aborted;
+                }
+
+                if (_state.ShutdownState != ShutdownState.Finished)
+                {
+                    _state.ShutdownState = ShutdownState.Pending;
+                }
+                else if (!callShutdown && !abortRead)
```

Comment:
This logic seems weird... you're only going to get to this case if ShutdownState is Finished, and then you set it to Finished again?

Reply:
may be leftover. I tried and failed to use the shutdown state. So at some point I added ShutdownDone and I would hope we can remove `ShutdownState` eventually completely. I just did not make it in this iteration.

Reply:
BTW I was talking about shutdown sequence with @nibanks and there are really two (at most) independent directions - one for reading and for writing. So we are done when both are done or were never present in case of unidirectional stream.

Reply:
You are right @geoffkizer. I made change to transition to Pending if not shutdown was attempted. The other states are Cancelled or ConnectionClosed. I'm not sure if we really need the state besides ConnectionClosed where we hand out error code. But I'm not ready to rip all that out at this moment.